[excel] (Comments) Creating a separate comments article (#1366)
OfficeDev/office-js-docs-pr: docs/excel/excel-add-ins-comments.md

---
title: Work with comments using the Excel JavaScript API (preview)
description: ''
ms.date: 10/15/2019
localization_priority: Normal
---

# Work with comments using the Excel JavaScript API (preview)

> [!NOTE]
> The comment APIs are currently available only in public preview. [!INCLUDE [Information about using preview APIs](../includes/using-excel-preview-apis.md)]

This article describes how to add, read, modify, and remove comments in a workbook with the Excel JavaScript API. You can learn more about the comment feature from the [Insert comments and notes in Excel](https://support.office.com/article/insert-comments-and-notes-in-excel-bdcc9f5d-38e2-45b4-9a92-0b2b5c7bf6f8) article.

> In the Excel JavaScript API, a comment is both the initial note and the connected threaded discussion. It is tied to an individual cell. Every comment can be replied to by people viewing the workbook (that have sufficient permissions). A [Comment](/javascript/api/excel/excel.comment) object stores those replies as [CommentReply](/javascript/api/excel/excel.commentreply) objects. You should consider a comment to be a thread and that a thread must have a special entry as the starting point.

Reviewer: ```suggestion
In the Excel JavaScript API, a comment is both the initial note and the connected threaded discussion. It is tied to an individual cell. Anyone viewing the workbook with sufficient permissions can reply to a comment. A [Comment](/javascript/api/excel/excel.comment) object stores those replies as [CommentReply](/javascript/api/excel/excel.commentreply) objects. You should consider a comment to be a thread and that a thread must have a special entry as the starting point.
```

Reviewer: Not sure abou thte "special entry" phrasing - makes it seem like initial note has to be something super important but YMMV

![An Excel comment, labelled "Comment" with two replies, labelled "Comment.replies[0]" and "Comment.replies[1].](../images/excel-comments.png)

Comments within a workbook are tracked by the `Workbook.comments` property. This includes comments created by users and also comments created by your add-in. The `Workbook.comments` property is a [CommentCollection](/javascript/api/excel/excel.commentcollection) object that contains a collection of [Comment](/javascript/api/excel/excel.comment) objects. Comments are also accessible at the [Worksheet](/javascript/api/excel/excel.worksheet) level. The samples in this article work with comments at the workbook level, but they can be easily modified to use the `Worksheet.comments` property.

## Add comments

> To add comments to a workbook, use the `CommentCollection.add` method, passing in the cell where the comment will be added, as either a string or [Range](/javascript/api/excel/excel.range) object, and the comment's text, as a string. Comments added by an add-in are attributed to the current user of that add-in.

Reviewer: Optionally could make this bullets, e.g. 
Use the method, like so:
-Pass in the cell 
- Add comment text as string

The following code sample adds a reply to the first comment in the workbook.

```js
Excel.run(function (context) {
    // Get the first comment added to the workbook.
    var comment = context.workbook.comments.getItemAt(0);
    comment.replies.add("Thanks for the reminder!");
    return context.sync();
});
```

## Edit comments

To edit a comment or comment reply, set its `Comment.content` property or `CommentReply.content` property.

```js
Excel.run(function (context) {
    // Edit the first comment in the workbook.
    var comment = context.workbook.comments.getItemAt(0);
    comment.content = "PLEASE add headers here.";
    return context.sync();
});
```

### Edit comment replies

To edit a comment reply, set its `CommentReply.content` property.

```js
Excel.run(function (context) {
    // Edit the first comment reply on the first comment in the workbook.
    var comment = context.workbook.comments.getItemAt(0);
    var reply = comment.replies.getItemAt(0);
    reply.content = "Never mind";
    return context.sync();
});
```

## Delete comments

> To delete a comment use the `Comment.delete` method. Deleting a comment also deletes all the replies associated with that comment.

Reviewer: ```suggestion
To delete a comment use the `Comment.delete` method. Deleting a comment also deletes the thread of replies associated with that comment.
```

```js
Excel.run(function (context) {
    // Delete the comment thread at A2 on the "MyWorksheet" worksheet.
    context.workbook.comments.getItemByCell("MyWorksheet!A2").delete();
    return context.sync();
});
```

### Delete comment replies

To delete a comment reply, use the `CommentReply.delete` method.

```js
Excel.run(function (context) {
    // Delete the first comment reply from this worksheet's first comment.
    var comment = context.workbook.comments.getItemAt(0);
    comment.replies.getItemAt(0).delete();
    return context.sync();
});
```

## Resolve comment threads

> A comment thread has a settable boolean value, `resolved`, to indicate if it is resolved. A value of `true` means the comment thread is resolved. A value of `false` means the comment thread is either new or reopened.

Reviewer: Maybe configurable instead of settable?

```js
Excel.run(function (context) {
    // Resolve the first comment thread in the workbook.
    context.workbook.comments.getItemAt(0).resolved = true;
    return context.sync();
});
```

Comment replies have a readonly `resolved` property. Its value is always equal to that of the rest of the thread.

## Comment metadata

Each comment contains metadata about its creation, such as the author and creation date. Comments created by your add-in are considered to be authored by the current user.

The following sample shows how to display the author's email, author's name, and creation date of a comment at **A2**.

```js
Excel.run(function (context) {
    // Get the comment at cell A2 in the "MyWorksheet" worksheet.
    var comment = context.workbook.comments.getItemByCell("MyWorksheet!A2");

    // Load and print the following values.
    comment.load(["authorEmail", "authorName", "creationDate"]);
    return context.sync().then(function () {
        console.log(`${comment.creationDate.toDateString()}: ${comment.authorName} (${comment.authorEmail})`);
    });
});
```

### Comment reply metadata

> Comment replies store the same types of metadata as the base comment.

Reviewer: ```suggestion
Comment replies store the same types of metadata as the initial comment.
```
